### PR TITLE
Improve LearningCircuit class and Regressor, add GD solver, add test file

### PR DESCRIPTION
**Pull request by Gruntrexpewrus**
Hi, it's Leonardo and this is a list of what I would like to pull:
1) In `skqulacs/circuit/circuit.py` I added a function to perform any multiqubit gate using qulacs base function.
2) In `skqulacs/qnn/solver` I added gradient descent method.
3) In `skqulacs/qnn/regressor` I added: `_func_grad` which computes the gradients of the circuit without the cost, I added `observables_str: List[str] = field(init=True, default_factory=list)` for the class and modified around to make it possible to use. The latter allows to arbitrary select the observable writing them as a string (e.g. "Z 3"). I think it was a bit more complex. Since is initialized as empty list, it does not cause problems in any test code.
4) In `skqulacs/test` I added `test_qnn_regressor_grad_descent.py` which in the style of the other tests shows that is possible to learn the sin (noisy or not) with batch gradient. Careful, it seems that stochastic gradient does not work, and it may be a simple theoretical reason. 

Feel free to contact me if there is any error in the codes or PR!

### CODE REVIEW

**Comment by Gruntrexpewrus** [comment]
There is a little bug related to the operators, I will fix.

File: skqulacs/circuit/circuit.py [hunk 1]
```diff
@@ -500,6 +507,35 @@ def _add_parametric_R_gate_inner(
 
         return parameter_id
 
+    def _add_multi_qubit_parametric_R_gate_inner(
```
Comment (forest1040) [inline skqulacs/circuit/circuit.py:510]: This function is not called. Why do you defined this function?
if it isn't nessesary, delete it.

File: skqulacs/qnn/regressor.py [hunk 2]
```diff
@@ -67,6 +68,20 @@ def __post_init__(self) -> None:
...
             )
+        if self.observables_str != []:
+            # add the right observables
+            for i in self.observables_str:
+                observable = Observable(self.n_qubit)
+                observable.add_operator(1.0, i)
+                self.observables.append(observable)
+        else:
+            # just add Zi for number of i outputs
+            for i in range(self.n_outputs):
```
Comment (forest1040) [inline skqulacs/qnn/regressor.py:79]: self.n_outputs is initialized 1 vlaue. But it's not so good. line 56
My suggetion is `for i in range(self.n_outputs):` change to `for i in range(1):`

File: skqulacs/qnn/regressor.py [hunk 3]
```diff
@@ -186,12 +195,35 @@ def _cost_func_grad(
         for h in range(len(x_scaled)):
             backobs = Observable(self.n_qubit)
             if self.n_outputs >= 2:
-                for i in range(self.n_outputs):
+                for i in self.observables_str:
                     backobs.add_operator(
-                        (-y_scaled[h][i] + mto[h][i]) / self.n_outputs, f"Z {i}"
+                        2 * (-y_scaled[h][i] + mto[h][i]) / self.n_outputs,
+                        i,  # I add a 2* as a derivative of the RMSE error
```
Comment (forest1040) [inline skqulacs/qnn/regressor.py:201]: Why do you add a 2* ?
this sentence `a derivative of the RMSE error`, which i can't understand.
Please explanation detail.

**Review by forest1040 — COMMENTED** [review]
Unit test error ocared. Please check it.
https://github.com/Qulacs-Osaka/scikit-qulacs/actions/runs/3580686474/jobs/6023019436

File: skqulacs/qnn/regressor.py (hunk 2 above)
Comment (Gruntrexpewrus) [reply skqulacs/qnn/regressor.py:79]: Hi, thank you for the comments :) I am trying to resolve this issue. The problem is that sometime we have tests with more observables (2 outputs), so I am trying to make possible to define also 2 observables instead of only 1

File: skqulacs/qnn/regressor.py (hunk 3 above)
Comment (Gruntrexpewrus) [reply skqulacs/qnn/regressor.py:201]: I think that since the RMSE is (-y_label+y_pred)^2, we need to account the derivative as 2*(-y_label+y_pred)dy_pred/dtheta. So that's why I multiplied for a 2. Is just a constant anyway.

File: skqulacs/circuit/circuit.py (hunk 1 above)
Comment (Gruntrexpewrus) [reply skqulacs/circuit/circuit.py:510]: This can be used to define 2qubit parametric gates. In the repository there was already something similar, but now is completed. Is not used in codes, but allows to recreate easily some ansatz with 2-qubit parametric gates.

Comment (Gruntrexpewrus) [reply skqulacs/circuit/circuit.py:510]: I actually use it in some personal codes DX

File: skqulacs/qnn/regressor.py (hunk 2 above)
Comment (forest1040) [reply skqulacs/qnn/regressor.py:79]: OK. Can you check `self.n_outputs` is `None`?

Comment (Gruntrexpewrus) [reply skqulacs/qnn/regressor.py:79]: I may have solved the issue, I will now pr again it seems to work locally now

**Comment by Gruntrexpewrus** [comment]
⚠️⚠️⚠️ There still is a bug related to the observables, I will update here when is solved. Maybe I will also modify the test file to show a circuit where I manually select the observables.

File: skqulacs/circuit/circuit.py (hunk 1 above)
Comment (forest1040) [reply skqulacs/circuit/circuit.py:510]: Can you integration `add_parametric_multi_Pauli_rotation_gate()` function?

Comment (Gruntrexpewrus) [reply skqulacs/circuit/circuit.py:510]: OK, I will insert another circuit in the test file and create an ansatz with the function.

File: skqulacs/qnn/regressor.py (hunk 3 above)
Comment (forest1040) [reply skqulacs/qnn/regressor.py]: we times 2 at qualcs backprop. because you don't have to add *2..
https://github.com/qulacs/qulacs/blob/main/src/vqcsim/parametric_circuit.cpp#L274

Comment (Gruntrexpewrus) [reply skqulacs/qnn/regressor.py]: Ok, Thank you for the explanation. I will remove the 2*

Comment (WATLE) [reply skqulacs/qnn/regressor.py]: Upon verification, qualcs backprop was irrelevant. The original code was wrong, and *2 was correct.  Thank you.

Comment (Gruntrexpewrus) [reply skqulacs/qnn/regressor.py]: Ok I understand, I will correct then! Thank you for the feedback

File: tests/test_qnn_regressor_grad_descent.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            y_batch = y_train[selected]
+            opt_loss, opt_params = qnn.fit(x_batch, y_batch, lr)
+            opt_loss = qnn.cost_func(opt_params, x_train, y_train)
+            losses.append(opt_loss)
+    x_test, y_test = generate_noisy_data(x_min, x_max, (num_x, 1), sine)
+    y_pred = qnn.predict(x_test)
+    return x_test, y_test, y_pred, losses
+
+
+def just_gradients(str_ob) -> Tuple[NDArray[np.float_]]:
```
Comment (forest1040) [inline tests/test_qnn_regressor_grad_descent.py:82]: Test code need start with test_xxx name.

**Comment by Gruntrexpewrus** [comment]
It seems that the PR passed all the checks 🥹

File: skqulacs/qnn/solver.py [hunk 5]
```diff
@@ -140,3 +140,27 @@ def run(
...
+@dataclass
+class Grad_Descent(Solver):
+    def run(
+        self,
+        cost_func: CostFunc,
+        jac: Jacobian,
+        theta: List[float],
+        x: NDArray[np.float_],
+        y: NDArray[np.float_],
+        lr: 0.1,
```
Comment (ikanago) [inline skqulacs/qnn/solver.py:154]: For simple gradient descent, learning rate is not changed in iterations. So the `lr` should be a member variable, I believe.
For example, [`SGD`](https://pytorch.org/docs/stable/generated/torch.optim.SGD.html#torch.optim.SGD) in PyTorch is initialized with the learning rate.

File: skqulacs/qnn/regressor.py [hunk 6]
```diff
@@ -49,10 +49,11 @@ class QNNRegressor:
     do_y_scale: bool = field(default=True)
     x_norm_range: float = field(default=1.0)
     y_norm_range: float = field(default=0.7)
+    observables_str: List[str] = field(init=True, default_factory=list)
```
Comment (ikanago) [inline skqulacs/qnn/regressor.py:52]: Could you explain an advantage of `observables_str`? I was not able to figure out.

**Review by ikanago — CHANGES_REQUESTED** [review]

File: skqulacs/qnn/regressor.py (hunk 6 above)
Comment (Gruntrexpewrus) [reply skqulacs/qnn/regressor.py:52]: I think that before you could not specify the observable when creating the Regressor, it was set as Z_i based in the i dimensions of the output. In this way you can declare an arbitrary observable (or set of observables)

File: tests/test_qnn_regressor_grad_descent.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        indexes = [idx for idx in range(num_x)]
+        for bc in range(num_x // batch_size):
+            selected = random.sample(indexes, batch_size)
+            for el in selected:
+                indexes.remove(el)
+            x_batch = x_train[selected]
+            y_batch = y_train[selected]
+            opt_loss, opt_params = qnn.fit(x_batch, y_batch, lr)
+            opt_loss = qnn.cost_func(opt_params, x_train, y_train)
+            losses.append(opt_loss)
```
[the remark this answers is not in the document]
Comment (Gruntrexpewrus) [reply tests/test_qnn_regressor_grad_descent.py:76]: As you said, I think for now is ok. If you want, I can work on something with PyTorch in another PR. If you mean a `Dataloader` class from scratch, I never did it outside PyTorch/TF, but may be another PR, and we need to select multiple datasets. I also think we should have separated files that generate datasets. I think also Qiskit Benchmarks has datasets files (which are not very efficient) but are easy to call around in any code.

File: skqulacs/qnn/regressor.py (hunk 6 above)
Comment (ikanago) [reply skqulacs/qnn/regressor.py:52]: I got it! But this is still complex...
How about passing `init=True` to `field` for `observables` so users can pass observables?
And move the initialization of default observables and initialize them only if `self.observable` is `[]`(when observables are not passed in constructor)?
```python
@dataclass
class QNNRegressor:
    # cut
    observables = field(init=True, defalut_factory=list)

    def __post_init__(self) -> None:
        # cut
        if len(self.observables) == 0:
            for i in range(n_outputs):
                observable = Observable(n_qubit)
                observable.add_operator(1.0, f"Z {i}")
                self.observables.append(observable)
```

Comment (Gruntrexpewrus) [reply skqulacs/qnn/regressor.py:52]: My philosophy was more that a user would prefer writing a string instead of looking at the documentation for how to define the operators.
But I think your solution is better, my problems are that I don't know then how to treat:
`        for h in range(len(x_scaled)):
            backobs = Observable(self.n_qubit)
            if self.n_outputs >= 2:
                for i in range(len(self.observables_str)):
                    backobs.add_operator(
                        2 * (-y_scaled[h][i] + mto[h][i]) / self.n_outputs,
                        self.observables_str[
                            i
                        ],`
and similar (in Regressor.py) because now I already have some files with Observables but I don't know which are they. Probably is just my lack of knowledge of qulacs operators 😂

Comment (Gruntrexpewrus) [reply skqulacs/qnn/regressor.py:52]: the current released version does `(-y_scaled[h][i] + mto[h][i]) / self.n_outputs, f"Z {i}"` that you see has very little freedom

File: tests/test_qnn_regressor_grad_descent.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (forest1040) [reply tests/test_qnn_regressor_grad_descent.py:76]: @ikanago I want `DataLoader` very much. Can you implement it ?

File: skqulacs/qnn/regressor.py (hunk 6 above)
Comment (ikanago) [reply skqulacs/qnn/regressor.py:52]: Ok, I agree with that passing string representations of an observable is easier.
The code you mention is the one in line 197-206, right? Currently we do not have any concise API in qulacs to construct an observable from existing one with a different coefficient. So we accept your code as a work around, and we are going to discuss how to deal with it in another issue in qulacs.

File: tests/test_qnn_regressor_grad_descent.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ikanago) [reply tests/test_qnn_regressor_grad_descent.py:76]: OK, I'll try in #256 !

File: skqulacs/qnn/regressor.py (hunk 6 above)
Comment (Gruntrexpewrus) [reply skqulacs/qnn/regressor.py:52]: I understand and I agree that it would be a good issue in qulacs. 
Yes, the code is from the position you mention :)

File: tests/test_qnn_regressor_grad_descent.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    x_test, y_test = generate_noisy_data(x_min, x_max, (num_x, 1), sine)
+    y_pred = qnn.predict(x_test)
```
[the remark this answers is not in the document]
Comment (Gruntrexpewrus) [reply tests/test_qnn_regressor_grad_descent.py:78]: I used MSE score now asserting at 0.045. Plain Gradient Descent and this ansatz can't do much better, but I check the plots and the learning is correct.

File: skqulacs/qnn/solver.py (hunk 5 above)
Comment (Gruntrexpewrus) [reply skqulacs/qnn/solver.py]: I used `lr: float = 0.1,` now, not sure if it's what you meant :)

Comment (Gruntrexpewrus) [reply skqulacs/qnn/solver.py]: ah, I also changed the name of `maxiter` in the fit function, because it can be `maxiter_or_lr`

**Comment by Gruntrexpewrus** [comment]
I think I followed your concerns @ikanago, let me know if I change other things 🤓

**Comment by forest1040** [comment]
Merged.
